external-crates/move/move-vm/profiler/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+    #[serde(rename(serialize = "type"))]
+    ty: String,
+    frame: u64,
+    at: u64,
+}
+
+#[cfg(debug_assertions)]
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Profile {
+    #[serde(rename(serialize = "type"))]
+    ty: String,
+    name: String,
+    unit: String,
+    start_value: u64,
+    end_value: u64,
+    events: Vec<Event>,
+}
+
+#[cfg(debug_assertions)]
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GasProfiler {
+    exporter: String,
+    name: String,
+    active_profile_index: u64,
+    #[serde(rename(serialize = "$schema"))]
+    schema: String,
+    shared: Shared,
+    profiles: Vec<Profile>,
+
+    #[serde(skip)]
+    pub start_gas: u64,
+    #[serde(skip)]
+    pub config: VMProfilerConfig,
+}
+
+#[cfg(debug_assertions)]
+impl GasProfiler {
+    pub fn init(config: &VMProfilerConfig, name: String, start_gas: u64) -> Self {
+        let mut prof = GasProfiler {
+            exporter: "speedscope@1.15.2".to_string(),
+            name: name.clone(),
+            active_profile_index: 0,
+            schema: "https://www.speedscope.app/file-format-schema.json".to_string(),
+            shared: Shared {
+                frames: vec![],
+                frame_table: BTreeMap::new(),
+            },
+            profiles: vec![Profile {
+                ty: "evented".to_string(),
+                name,
+                unit: "none".to_string(),
+                start_value: 0,
+                end_value: 0,
+                events: vec![],
+            }],
+            start_gas,
+            config: config.clone(),
+        };
+        profile_open_frame_impl!(Some(&mut prof), "root".to_string(), start_gas);
+        prof
+    }
+
+    pub fn init_default_cfg(name: String, start_gas: u64) -> Self {
+        Self::init(&VMProfilerConfig::default(), name, start_gas)
+    }
+
+    fn profile_name(&self) -> String {
+        self.name.clone()
+    }
+
+    pub fn short_name(s: &String) -> String {
+        s.split("::").last().unwrap_or(s).to_string()
+    }
+
+    fn is_metered(&self) -> bool {
+        (self.profiles[0].end_value != 0) && (self.start_gas != 0)
+    }
+
+    fn start_gas(&self) -> u64 {
+        self.start_gas
+    }
+
+    fn add_frame(&mut self, frame_name: String) -> u64 {
+        *self
+            .shared
+            .frame_table
+            .entry(frame_name.clone())
+            .or_insert({
+                let val = self.shared.frames.len() as u64;
+                self.shared.frames.push(FrameName { name: frame_name });
+                val as usize
+            }) as u64
+    }
+
+    pub fn open_frame(&mut self, frame_name: String, gas_start: u64) {
+        if !*PROFILER_ENABLED || self.start_gas == 0 {
+            return;
+        }
+
+        let frame_idx = self.add_frame(frame_name.clone());
+        let start = self.start_gas();
+
+        self.profiles[0].events.push(Event {
+            ty: "O".to_string(),
```

Comment:
What are these types here? With `"O"` and `"C"`?

Reply:
Open and Close.
Thats just how the format represents them.

Reply:
Might want to do a constant/comment or someway to make this transparent, just for future thought